contracts/ccip/ccip/sources/onramp.move

```diff
@@ -0,0 +1,1124 @@
+        header: RampMessageHeader {
+            message_id,
+            source_chain_selector,
+            dest_chain_selector,
+            sequence_number,
+            nonce,
+        },
+        sender,
+        data,
+        receiver,
+        extra_args,
+        fee_token,
+        fee_token_amount,
+        fee_value_juels: 0, // Not used in hashing
+        token_amounts,
+    };
+
+    calculate_message_hash_internal(&message, metadata_hash)
+}
+
+public fun calculate_metadata_hash(
+    source_chain_selector: u64,
+    dest_chain_selector: u64,
+    on_ramp_address: address,
+): vector<u8> {
+    let mut packed = vector[];
+    eth_abi::encode_right_padded_bytes32(
+        &mut packed,
+        hash::keccak256(&b"Sui2AnyMessageHashV1"),
+    );
+    eth_abi::encode_u64(&mut packed, source_chain_selector);
+    eth_abi::encode_u64(&mut packed, dest_chain_selector);
+    eth_abi::encode_address(&mut packed, on_ramp_address);
+    hash::keccak256(&packed)
+}
+
+fun calculate_message_hash_internal(
+    message: &Sui2AnyRampMessage,
+    metadata_hash: vector<u8>,
+): vector<u8> {
+    let mut outer_hash = vector[];
+    eth_abi::encode_right_padded_bytes32(&mut outer_hash, merkle_proof::leaf_domain_separator());
+    eth_abi::encode_right_padded_bytes32(&mut outer_hash, metadata_hash);
+
+    let mut inner_hash = vector[];
+    eth_abi::encode_address(&mut inner_hash, message.sender);
+    eth_abi::encode_u64(&mut inner_hash, message.header.sequence_number);
+    eth_abi::encode_u64(&mut inner_hash, message.header.nonce);
+    eth_abi::encode_address(&mut inner_hash, message.fee_token);
+    eth_abi::encode_u64(&mut inner_hash, message.fee_token_amount);
+    eth_abi::encode_right_padded_bytes32(&mut outer_hash, hash::keccak256(&inner_hash));
+
+    eth_abi::encode_right_padded_bytes32(
+        &mut outer_hash,
+        hash::keccak256(&message.receiver),
+    );
+    eth_abi::encode_right_padded_bytes32(&mut outer_hash, hash::keccak256(&message.data));
+
+    let mut token_hash = vector[];
+    eth_abi::encode_u256(&mut token_hash, message.token_amounts.length() as u256);
+    message.token_amounts.do_ref!(|token_transfer| {
+        let token_transfer: &Sui2AnyTokenTransfer = token_transfer;
+        eth_abi::encode_address(
+            &mut token_hash,
+            token_transfer.source_pool_address,
+        );
+        eth_abi::encode_bytes(&mut token_hash, token_transfer.dest_token_address);
+        eth_abi::encode_bytes(&mut token_hash, token_transfer.extra_data);
+        eth_abi::encode_u64(&mut token_hash, token_transfer.amount);
+        eth_abi::encode_bytes(&mut token_hash, token_transfer.dest_exec_data);
+    });
+    eth_abi::encode_right_padded_bytes32(&mut outer_hash, hash::keccak256(&token_hash));
+
+    eth_abi::encode_right_padded_bytes32(
+        &mut outer_hash,
+        hash::keccak256(&message.extra_args),
+    );
+
+    hash::keccak256(&outer_hash)
+}
+
+public fun ccip_send<T>(
```

Comment:
for a function with both `CCIPObjectRef` and `OnRampState` or `CCIPObjectRef` and `OffRampState`, we can probably have a rule to say that we will always put this curse on `CCIPObjectRef` to minimize confusion.